web/elm/src/Views/CommentBar.elm

```diff
@@ -0,0 +1,305 @@
+                        [ style "width" "60px"
+                        , style "display" "flex"
+                        , style "justify-content" "flex-end"
+                        ]
+                        (case model.state of
+                            Viewing _ ->
+                                [ editButton model state ]
+
+                            Editing _ ->
+                                [ saveButton model state ]
+
+                            Saving _ ->
+                                [ loadingButton ]
+                        )
+                    ]
+
+                else
+                    [ Html.pre
+                        Styles.commentBarText
+                        [ Html.text (getContent model) ]
+                    ]
+               )
+        )
+
+
+commentSetCallback : ( String, Result Http.Error () ) -> Model -> Model
+commentSetCallback ( content, result ) model =
+    let
+        updatedState =
+            case ( result, model.state ) of
+                ( Ok (), Viewing _ ) ->
+                    Viewing content
+
+                ( Ok (), Editing state ) ->
+                    Editing { state | cached = content }
+
+                ( Ok (), Saving _ ) ->
+                    Viewing content
+
+                ( Err _, Saving state ) ->
+                    Editing state
+
+                ( _, state ) ->
+                    state
+    in
+    { model | state = updatedState }
+
+
+update : Message.Message -> Model -> ( Model, List Effect )
+update msg model =
+    case msg of
+        Message.Click (Message.CommentBarButton kind id) ->
+            if model.id == id then
+                case kind of
+                    Message.Edit ->
+                        case model.state of
+                            Viewing content ->
+                                ( { model | state = Editing { content = content, cached = content } }, [] )
+
+                            _ ->
+                                ( model, [] )
+
+                    Message.Save ->
+                        case model.state of
+                            Editing state ->
+                                if state.content /= state.cached then
+                                    ( { model | state = Saving state }, [ Save state.content ] )
+
+                                else
+                                    ( { model | state = Viewing state.content }, [] )
+
+                            _ ->
+                                ( model, [] )
+
+            else
+                ( model, [] )
+
+        Message.EditCommentBar id content ->
+            if model.id == id then
+                ( { model
+                    | state =
+                        case model.state of
+                            Editing state ->
+                                Editing { state | content = content }
+
+                            state ->
+                                state
+                  }
+                , []
```

Comment:
When we click on the edit button, we also want to `Focus` on the textarea. Maybe it's easier for `CommentBar.update` to return a `List Effects.Effect` so we don't need these intermediate `CommentBar`-local `Effects` that just get mapped to the "actual" `Effects.Effect`

Reply:
Was already working on this after my last commit 😛

I'd originally had issues passing a `List Effects.Effect` out of `CommentBar.update`, because I couldn't figure out how to pass a lambda in to generate the correct "save" effect as it can differ depending on where the comment bar is used (eg. builds use `SetBuildComment` whereas pin comments use `SetPinComment`). I took another look at it in the latest commit and have created another type alias (the poorly named, `UpdateState`) that is passed into `CommentBar.update` with the lambda to call.

https://github.com/concourse/concourse/blob/435e3987bb10b67888a3766bad6f364816258490/web/elm/src/Views/CommentBar.elm#L43-L45

https://github.com/concourse/concourse/blob/435e3987bb10b67888a3766bad6f364816258490/web/elm/src/Views/CommentBar.elm#L319